### PR TITLE
perf(ripple): do not register events if ripples are disabled initially

### LINKED ISSUE #8854: [MatCheckbox] nested ripple registers events even if disableRipple is set.
#### Bug, feature request, or proposal:
When using MatCheckbox, MatRipple and RippleRenderer is instantiated and registers events even if disableRipple is set on the checkbox.

#### What is the expected behavior?
If disableRipple is set, MatRipple and RippleRenderer should not be instantiated, or at the very least the events should not be registered during the initial render.

#### What is the current behavior?
There are 10 events registered for each checkbox, slowing down the render of the checkbox.

#### What are the steps to reproduce?
http://embed.plnkr.co/PZ0P3EerE2aeJPR4Kasa/

When clicking on the button to view Material Checkboxes, 30-45% of the time is spent on adding event listeners.

#### What is the use-case or motivation for changing an existing behavior?
This is visible when using a virtual rendering component to show a large number of items with checkboxes and scrolling, or when rendering the full list as in here. The result is slow rendering or janky scrolling.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
All versions are affected. Tested with Angular 2 with Chrome 62 on MacOS.

### PR DESCRIPTION
**Pull request by DevVersion**
* No longer registers trigger event listeners if the ripples are disabled initially.

Fixes #8854.

### CODE REVIEW

File: src/lib/core/ripple/ripple.ts [hunk 1]
```diff
@@ -86,36 +77,71 @@ export class MatRipple implements OnChanges, OnDestroy {
...
+   */
+  @Input('matRippleTrigger')
+  get trigger() { return this._trigger; }
+  set trigger(trigger: HTMLElement | null) {
+    this._trigger = trigger;
+
+    // The events for a trigger element shouldn't be registered if ripples are disabled.
+    // As soon as the ripples are enabled again, the events for the current trigger will be added.
+    if (!this.disabled && this._isInitialized) {
+      this._rippleRenderer.setTriggerElement(trigger);
```
Comment (jelbourn) [inline src/lib/core/ripple/ripple.ts]: It feels strange to call `setTriggerElement` when you really just want one of the _side-effects_ of `setTriggerElement`. Would it be easier to follow if there was a function to specifically add/remove the event listeners?

Comment (DevVersion) [reply src/lib/core/ripple/ripple.ts]: What do you mean with side-effects? Maybe the function name is just not good and it should be called `registerTriggerEvents()` or something like that.

Comment (jelbourn) [reply src/lib/core/ripple/ripple.ts]: Based on the name, the primary purpose of `setTriggerElement` would be the line
```ts
 this._triggerElement = element;
```

The event stuff is incidental to updating the trigger element. It be a bit better as something like
```ts
  /** Sets the trigger element and registers the mouse events. */
  setTriggerElement(element: HTMLElement | null) {
    if (this._triggerElement === element) {
      return;
    }

    this._removeTriggerListeners();
    this._triggerElement = element;
    this._addTriggerListeners();
  }
```

Then other call site that don't need to update to a new element just call `_removeTriggerListeners` and `_addTriggerListeners`

Comment (DevVersion) [reply src/lib/core/ripple/ripple.ts]: I see what you mean, but the actual event registration should be the primary purpose of this function (or part in the `RippleRenderer`)

In that case I would just call the method `_setupTriggerEvents()` and keep the implementation as it is. The `_triggerElement` variable is just there for the event un-registration, and the real `trigger` element is stored in the `MatRipple` class. Does that work for you?

File: src/lib/tabs/tab-nav-bar/tab-nav-bar.ts [hunk 2]
```diff
@@ -229,6 +228,7 @@ export class MatTabLink extends _MatTabLinkMixinBase
     // Manually create a ripple instance that uses the tab link element as trigger element.
     // Notice that the lifecycle hooks for the ripple config won't be called anymore.
     this._tabLinkRipple = new MatRipple(_elementRef, ngZone, platform, globalOptions);
+    this._tabLinkRipple.ngOnInit();
```
Comment (crisbeto) [inline src/lib/tabs/tab-nav-bar/tab-nav-bar.ts]: It feels weird having to call the Angular lifecycle hook manually. Shouldn't the renderer support attaching to arbitrary elements?

File: src/lib/core/ripple/ripple.ts [hunk 3]
```diff
@@ -86,36 +77,71 @@ export class MatRipple implements OnChanges, OnDestroy {
...
+  ngOnInit() {
+    this._isInitialized = true;
 
-    this._updateRippleRenderer();
+    // The trigger element might have been set before the `ngOnInit` lifecycle hook, but the
+    // events haven't been registered yet. This has been deferred to the `ngOnInit` lifecycle hook,
+    // because the events should be only registered if the ripples aren't disabled initially.
+    // Also if there is no trigger element, that has been specified explicitly through the input,
+    // the default trigger element will be the directive's host element.
+    this.trigger = this._trigger === undefined ? this._hostElement : this.trigger;
```
Comment (crisbeto) [inline src/lib/core/ripple/ripple.ts]: Doing this just so the getter runs is somewhat side-effect-ey. Why not have a private method that does the attachment both here and insider the setter?

File: src/lib/core/ripple/ripple.ts (hunk 1 above)
Comment (crisbeto) [reply src/lib/core/ripple/ripple.ts]: Agreed that at least the naming should change to something involving "events".

File: src/lib/tabs/tab-nav-bar/tab-nav-bar.ts (hunk 2 above)
Comment (DevVersion) [reply src/lib/tabs/tab-nav-bar/tab-nav-bar.ts]: Addressed the solution we discussed on Slack. I still kind of feel that the first commit is more clean and it's very uncommon that we want to use the `RippleRenderer` without the `MatRipple` directive.

**Comment by DevVersion** [comment]
@jelbourn Done. Please apply merge ready if you think it's good to go.

**Pull request merged**